imap_processing/cli.py

```diff
@@ -104,37 +148,101 @@ def _validate_args(args):
 
 
 class ProcessInstrument(ABC):
     """An abstract base class containing a method to process an instrument.
 
     Attributes
     ----------
-    level : str
+    data_level : str
         The data level to process (e.g. ``l1a``)
+    dependency_str : str
+        A string representation of the dependencies for the instrument in the
+        format: "[{
+            'instrument': 'mag',
+            'data_level': 'l0',
+            'descriptor': 'sci',
+            'version': 'v00-01',
+            'start_date': '20231212',
+            'end_date': '20231212'
+        }]"
+    start_date : str
+        The start date for the output data. Format: YYYYMMDD
+    end_date : str
+        The end date for the output data. Format: YYYYMMDD
+    version : str
+        The version of the data. Format: vxx-xx
+    upload_to_sdc : bool
+        A flag indicating whether to upload the output file to the SDC.
     """
 
-    def __init__(self, level, file_path):
-        self.level = level
-        self.file_path = file_path
+    def __init__(
+        self,
+        data_level: str,
+        dependency_str: str,
+        start_date: str,
+        end_date: str,
+        version: str,
+        upload_to_sdc: bool,
+    ) -> None:
+        self.data_level = data_level
+
+        # Convert string into a dictionary
+        self.dependencies = loads(dependency_str.replace("'", '"'))
+
+        self.start_date = start_date
+        self.end_date = end_date
+        if not end_date:
+            self.end_date = start_date
+            print(f"Setting end time to start time: {start_date}")
+
+        self.version = version
+        self.upload_to_sdc = upload_to_sdc
+
+    def download_dependencies(self):
+        """Download the dependencies for the instrument.
+
+        Returns
+        -------
+        file_list: list[str]
+            A list of file paths to the downloaded dependencies.
+        """
+        file_list = []
+        for dependency in self.dependencies:
+            try:
+                # TODO: Validate dep dict
+                # TODO: determine what dependency information is optional
+                # TODO: Add in timestamps and descriptor to query
+                return_query = imap_data_access.query(
+                    instrument=dependency["instrument"],
+                    data_level=dependency["data_level"],
+                    version=dependency["version"],
+                )
+            except HTTPError as e:
+                raise ValueError(f"Unable to download files from {dependency}") from e
+
+            if not return_query:
+                raise FileNotFoundError(
+                    f"File not found for required dependency "
+                    f"{dependency} while attempting to create file."
+                    f"This should never occur "
+                    f"in normal processing."
+                )
+
+            file_list.append(imap_data_access.download(return_query[0]["file_path"]))
```

Comment:
Above you have the "Unable to download files" message. Should this be moved up into that try/except block so we also catch the HTTPError there?

How is return_query sorted? I would guess we want the latest file `[-1]` or all of them even by iterating over the list returned?

Reply:
Right now, we would expect only one file to return for a given dependency query. That's not the case right now, but Tenzin needs to update the dependency management in batch_starter first.